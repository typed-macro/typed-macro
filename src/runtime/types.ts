import { CallExpression } from '@babel/types'
import { NodePath } from '@babel/traverse'
import { PathHelper } from '@/helper/path'
import { StateHelper } from '@/helper/state'
import { TransformHelper } from '@/helper/transform'
import { Babel } from '@/helper/babel'

export type MacroContext = {
  /**
   * The file path of the module currently being handled.
   */
  filepath: string
  /**
   * The original source code of the module currently being handled.
   */
  code: string
  /**
   * The arguments nodes of the call-macro expression currently being handled.
   */
  args: CallExpression['arguments']
  /*
   * The NodePath of the call-macro expression currently being handled.
   */
  path: NodePath
  /*
   * Whether in ssr mode.
   */
  ssr: boolean
}

export type MacroHelper = TransformHelper & PathHelper & StateHelper

export type MacroHandler = (
  /**
   * Context about the macro caller.
   */
  ctx: MacroContext,
  /**
   * Babel tools to help operate node paths and nodes.
   */
  babel: Readonly<Babel>,
  /**
   * Wrappers upon babel tools to simplify operations.
   */
  helper: Readonly<MacroHelper>
) => void

export type Macro = {
  name: string
  apply: MacroHandler
}

export type NamespacedMacros = { [namespace: string]: Macro[] }

export type NamespacedModules = { [namespace: string]: string }

export type MacroWithType = Macro & {
  __types: string
}

export function isMacroWithType(o: unknown): o is MacroWithType {
  return !!(o as MacroWithType).__types
}

export type Exportable = ({ macros: Macro[] } | { code: string }) & {
  customTypes?: string
}

export type NamespacedExportable = { [namespace: string]: Exportable }

export type NormalizedExports = {
  macros: NamespacedMacros
  modules: NamespacedModules
  types: NamespacedTypes
}

export function normalizeExports(
  exports: NamespacedExportable
): NormalizedExports {
  const macros: NamespacedMacros = Object.create(null)
  const modules: NamespacedModules = Object.create(null)
  const types: NamespacedTypes = Object.create(null)
  Object.keys(exports).forEach((ns) => {
    const item = exports[ns]
    types[ns] = {
      moduleScope: item.customTypes || '',
      macroScope: [],
    }
    if ('code' in item) {
      modules[ns] = item.code
    } else {
      macros[ns] = item.macros.map((m) => ({ name: m.name, apply: m.apply }))
      types[ns].macroScope = item.macros.map((m) => {
        if (!isMacroWithType(m)) throw new Error(`${m.name} is not a macro`)
        return m.__types
      })
    }
  })
  return {
    macros,
    modules,
    types,
  }
}

export type NamespacedTypes = {
  [namespace: string]: {
    moduleScope: string
    macroScope: string[]
  }
}

export function renderTypes(types: NamespacedTypes) {
  const namespaces = Object.keys(types)
  return namespaces
    .map((ns) => {
      const item = types[ns]
      return `declare module '${ns}' {
${[item.moduleScope, item.macroScope.join('\n')].filter((t) => !!t).join('\n')}
}`
    })
    .join('\n')
}
