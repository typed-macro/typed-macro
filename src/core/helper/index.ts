export { BABEL_TOOLS, Babel } from './babel'
export { ImportOption } from './import'
export { getPathHelper, PathHelper } from './path'
export { getStateHelper, StateHelper } from './state'
export { getTransformHelper, TransformHelper } from './transform'
