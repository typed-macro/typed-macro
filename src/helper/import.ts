import {
  ImportDeclaration,
  isIdentifier,
  isImportDefaultSpecifier,
  isImportNamespaceSpecifier,
  isImportSpecifier,
} from '@babel/types'

/**
 * ```typescript
 * import 'moduleName'
 * { moduleName: string }
 *
 * import defaultName from 'moduleName'
 * { defaultName: string; moduleName: string }
 *
 * import { exportName } from 'moduleName'
 * { exportName: string; moduleName: string }
 *
 * import { exportName as localName } from 'moduleName'
 * { localName: string; exportName: string; moduleName: string }
 *
 * import * as namespaceName from 'moduleName'
 * { namespaceName: string; moduleName: string }
 * ```
 */
export type ImportOption =
  | { moduleName: string }
  | { defaultName: string; moduleName: string }
  | { localName: string; exportName: string; moduleName: string }
  | { exportName: string; moduleName: string }
  | { namespaceName: string; moduleName: string }

export const renderImportStmt = (imp: ImportOption) =>
  'defaultName' in imp
    ? `import ${imp.defaultName} from '${imp.moduleName}'`
    : 'localName' in imp
    ? `import { ${imp.exportName} as ${imp.localName} } from '${imp.moduleName}'`
    : 'exportName' in imp
    ? `import { ${imp.exportName} } from '${imp.moduleName}'`
    : 'namespaceName' in imp
    ? `import * as ${imp.namespaceName} from '${imp.moduleName}'`
    : `import '${imp.moduleName}'`

export function matchImportStmt(
  imp: ImportOption,
  node: ImportDeclaration,
  loose = true
) {
  if (node.source.value !== imp.moduleName) return false
  if ('defaultName' in imp) {
    // import defaultName from 'moduleName'
    for (const s of node.specifiers) {
      if (isImportDefaultSpecifier(s) && s.local.name === imp.defaultName)
        return true
    }
  } else if ('localName' in imp) {
    // import { exportName as localName } from 'moduleName'
    for (const s of node.specifiers) {
      if (
        isImportSpecifier(s) &&
        s.local.name === imp.localName &&
        isIdentifier(s.imported) &&
        s.imported.name === imp.exportName
      )
        return true
    }
  } else if ('exportName' in imp) {
    // import { exportName } from 'moduleName'
    for (const s of node.specifiers) {
      if (
        isImportSpecifier(s) &&
        isIdentifier(s.imported) &&
        s.imported.name === imp.exportName &&
        (loose || s.local.name === imp.exportName)
      )
        return true
    }
  } else if ('namespaceName' in imp) {
    // import * as defaultName from 'moduleName'
    for (const s of node.specifiers) {
      if (isImportNamespaceSpecifier(s) && s.local.name === imp.namespaceName)
        return true
    }
  } else {
    // import 'moduleName'
    if (loose || node.specifiers.length === 0) return true
  }
  return false
}
