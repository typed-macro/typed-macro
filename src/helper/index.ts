export * from './macro'
export * from './babel'
export type { ImportOption } from './import'
export type { DevServerHelper } from './server'
