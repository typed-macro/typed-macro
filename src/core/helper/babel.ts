import template from '@babel/template'
import traverse from '@babel/traverse'
import * as types from '@babel/types'
import { parse, parseExpression } from '@babel/parser'
import generate from '@babel/generator'

export type Babel = {
  /**
   * From @babel/template.
   * Help generate nodes from code string.
   *
   * e.g.
   * ```typescript
   * template.statement.ast('const a = 1') // generate a statement node
   * ```
   */
  template: typeof template

  /**
   * From @babel/traverse.
   * Help traverse from a node.
   *
   * e.g.
   * ```typescript
   * traverse(someNode, {
   *   CallExpression(path) {
   *     ...
   *   }
   * })
   * ```
   */
  traverse: typeof traverse

  /**
   * From @babel/types.
   * Help check node type and create nodes of specific type.
   *
   * e.g.
   * ```typescript
   *  types.isIdentifier(someNode) // check is identifier
   *  types.identifier('hello') // create identifier node
   * ```
   */
  types: typeof types

  /**
   * From @babel/parser.
   * Help parse code string to ast.
   *
   * e.g.
   * ```typescript
   *  parse('const a = 1')
   * ```
   */
  parse: typeof parse

  /**
   * From @babel/parser.
   * Help parse code string to expression.
   *
   * e.g.
   * ```typescript
   *  parseExpression('1 + 2')
   * ```
   */
  parseExpression: typeof parseExpression

  /**
   * From @babel/generator.
   * Help generate code string from ast.
   *
   * e.g.
   * ```typescript
   *  generate(someNode)
   * ```
   */
  generate: typeof generate
}

export const BABEL_TOOLS: Babel = Object.freeze({
  template,
  traverse,
  types,
  parse,
  parseExpression,
  generate,
})
