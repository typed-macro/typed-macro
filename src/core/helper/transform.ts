import { Node, NodePath } from '@babel/traverse'
import template from '@babel/template'
import { File, ImportDeclaration, Program } from '@babel/types'
import { ImportOption, matchImportStmt, renderImportStmt } from './import'
import { findProgramPath } from '@/common'

export type TransformHelper = {
  /**
   * Prepend import statements to program.
   * @param imports an import or an array of imports
   * @param program node path of the target program for prepending import statements.
   * use the one currently being handled by default.
   * @return node path of the last inserted import statement,
   *   or undefined if imports param is array and length === 0
   */
  prependImports: (
    imports: ImportOption[] | ImportOption,
    program?: NodePath<Program>
  ) => NodePath<ImportDeclaration> | undefined

  /**
   * Append import statements to the last import statement of the program.
   * @param imports an import or an array of imports
   * @param program node path of the target program for prepending import statements.
   * use the one currently being handled by default.
   * @return node path of the last inserted import statement
   *   or undefined if imports param is array and length === 0
   */
  appendImports: (
    imports: ImportOption[] | ImportOption,
    program?: NodePath<Program>
  ) => NodePath<ImportDeclaration> | undefined

  /**
   * Find an import statement that has been in the target program already.
   * Note that findImported() will not resolve extension for moduleName automatically.
   * @param imports an import
   * @param loose match loosely, defaults to true, e.g:
   * `import { a as _a } from 'a'` will match `{ moduleName: 'a' }` and `{ moduleName: 'a', exportName: 'a' }`
   * @param program node path of the target program. use the one currently being handled by default.
   */
  findImported: (
    imports: ImportOption,
    loose?: boolean,
    program?: NodePath<Program>
  ) => NodePath<ImportDeclaration> | undefined

  /**
   * Check if an import statement has been in the target program already.
   * Note that hasImported() will not resolve extension for moduleName automatically.
   * @param imports an import
   * @param loose match loosely, defaults to true, e.g:
   * `import { a as _a } from 'a'` will match `{ moduleName: 'a' }` and `{ moduleName: 'a', exportName: 'a' }`
   * @param program node path of the target program. use the one currently being handled by default.
   */
  hasImported: (
    imports: ImportOption,
    loose?: boolean,
    program?: NodePath<Program>
  ) => boolean

  /**
   * Prepend any node to the target program.
   * @param nodes any node or an array of nodes
   * @param program node path of the target program. use the one currently being handled by default.
   * @return node path of the last inserted node
   */
  prependToBody: <Nodes extends Node | readonly Node[]>(
    nodes: Nodes,
    program?: NodePath<Program>
  ) => NodePath

  /**
   * Append any node to the target program.
   * @param nodes any node or an array of nodes
   * @param program node path of the target program. use the one currently being handled by default.
   * @return node path of the last inserted node
   */
  appendToBody: <Nodes extends Node | readonly Node[]>(
    nodes: Nodes,
    program?: NodePath<Program>
  ) => NodePath

  /**
   * Get current program or the program of provided node.
   * @param node any node path
   */
  getProgram: (node?: NodePath) => NodePath<Program>
}

export function getTransformHelper(ast: File): TransformHelper {
  const thisProgram = findProgramPath(ast)

  const findImported: TransformHelper['findImported'] = (
    imp,
    loose = true,
    program = thisProgram
  ) => {
    // An import declaration can only be used in top-level.
    for (const path of program.get('body') as NodePath[]) {
      if (!path.isImportDeclaration()) continue
      if (matchImportStmt(imp, path.node, loose)) return path
    }
  }

  const hasImported: TransformHelper['hasImported'] = (
    imp,
    loose = true,
    program = thisProgram
  ) => {
    return findImported(imp, loose, program) !== undefined
  }

  // returns
  // - array: generated import statements
  // - path: all import statements are duplicated, returns the node path of the last one
  // - undefined: none import statement provided
  function normalizeImports(
    imports: ImportOption | ImportOption[],
    program: NodePath<Program>
  ) {
    if (Array.isArray(imports)) {
      if (imports.length === 0) return
    } else {
      imports = [imports]
    }
    // remove duplicated
    const toBeImported = imports.filter(
      (imp) => !hasImported(imp, false, program)
    )
    // all import statements are duplicated, returns the node path of the last one
    if (!toBeImported.length)
      return findImported(imports[imports.length - 1], false, program)!
    return template.statements.ast(
      Array.from(
        new Set(toBeImported.map((imp) => renderImportStmt(imp)))
      ).join('; ')
    ) as ImportDeclaration[]
  }

  const prependImports: TransformHelper['prependImports'] = (
    imports,
    program = thisProgram
  ) => {
    const toBeImported = normalizeImports(imports, program)
    if (!toBeImported) return
    if (!Array.isArray(toBeImported)) return toBeImported
    const firstImport = (program.get('body') as NodePath[]).filter((p) =>
      p.isImportDeclaration()
    )[0]
    return (
      firstImport
        ? firstImport.insertBefore(toBeImported)
        : program.unshiftContainer('body', toBeImported)
    ).pop() as NodePath<ImportDeclaration>
  }

  const appendImports: TransformHelper['appendImports'] = (
    imports,
    program = thisProgram
  ) => {
    const toBeImported = normalizeImports(imports, program)
    if (!toBeImported) return
    if (!Array.isArray(toBeImported)) return toBeImported
    const lastImport = (program.get('body') as NodePath[])
      .filter((p) => p.isImportDeclaration())
      .pop()
    return (
      lastImport
        ? lastImport.insertAfter(toBeImported)
        : program.unshiftContainer('body', toBeImported)
    ).pop() as NodePath<ImportDeclaration>
  }

  const prependToBody: TransformHelper['prependToBody'] = (
    nodes,
    program = thisProgram
  ) => {
    return program.unshiftContainer('body', nodes).pop() as NodePath
  }

  const appendToBody: TransformHelper['appendToBody'] = (
    nodes,
    program = thisProgram
  ) => {
    return program.pushContainer('body', nodes).pop() as NodePath
  }

  const getProgram: TransformHelper['getProgram'] = (node) => {
    if (!node) return thisProgram
    return node.findParent((p) => p.isProgram()) as NodePath<Program>
  }

  return {
    findImported,
    hasImported,
    prependImports,
    appendImports,
    prependToBody,
    appendToBody,
    getProgram,
  }
}
