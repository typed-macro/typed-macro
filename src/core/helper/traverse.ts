import {
  CallExpression,
  File,
  isIdentifier,
  isImportDefaultSpecifier,
  isImportNamespaceSpecifier,
  isMemberExpression,
  isStringLiteral,
  Program,
} from '@babel/types'
import traverse, { NodePath, Scope } from '@babel/traverse'
import { ImportOption, matchImportStmt } from '@/core/helper/import'
import { Macro } from '@/core/macro'
import { NamespacedMacros } from '@/core/exports'
import template from '@babel/template'

export function findProgramPath(ast: File) {
  let path: NodePath<Program>
  traverse(ast, {
    Program(p) {
      path = p
      p.stop()
    },
  })
  return path!
}

export function findImported(
  program: NodePath<Program>,
  imp: ImportOption,
  loose = true
) {
  // An import declaration can only be used in top-level.
  for (const path of program.get('body') as NodePath[]) {
    if (!path.isImportDeclaration()) continue
    if (matchImportStmt(imp, path.node, loose)) return path
  }
}

export type ImportedAsNS = {
  importedAsNamespace: true
  namespace: string
  local: string
  macros: Macro[]
}

export type ImportedAsNamed = {
  importedAsNamespace: false
  namespace: string
  local: string
  // not always macro
  macro?: Macro
}

export type ImportedMacro = ImportedAsNS | ImportedAsNamed

export function findImportedMacros(
  ast: File,
  macros: NamespacedMacros,
  keepImportStmt = false
) {
  const importedMacros: ImportedMacro[] = []
  const namespaces = Object.keys(macros)

  traverse(ast, {
    ImportDeclaration(path) {
      const ns = path.node.source.value
      if (!namespaces.includes(ns)) return

      // case - import 'a'
      if (!path.node.specifiers.length) return

      path.node.specifiers.forEach((s) => {
        if (isImportDefaultSpecifier(s) || isImportNamespaceSpecifier(s)) {
          // case - import a from 'a'
          // or
          // case - import * as a from 'a'
          importedMacros.push({
            importedAsNamespace: true,
            namespace: path.node.source.value,
            local: s.local.name,
            macros: macros[ns],
          })
        } else {
          // case - import { a as a } from 'a'
          const exportName = isIdentifier(s.imported)
            ? s.imported.name
            : s.imported.value

          importedMacros.push({
            importedAsNamespace: false,
            namespace: path.node.source.value,
            local: s.local.name,
            macro: macros[ns].find((m) => m.name === exportName),
          })
        }
      })
      keepImportStmt
        ? path.replaceWith(
            template.statement.ast(`import '${path.node.source.value}'`)
          )
        : path.remove()
    },
  })

  // update scope so that macros don't have bindings now
  findProgramPath(ast).scope.crawl()

  return importedMacros
}

function isLocalDefined(scope: Scope, name: string) {
  // local-defined identifiers must have bindings,
  // macros don't have bindings,
  // but no bindings doesn't mean is a macro,
  // e.g. console.log(), `console` has no bindings too.
  return scope.getBinding(name)
}

// find macros to be applied from call expr, returns a
// - Macro: the macro to be applied
// - string: is a macro call but no macro found
// - undefined: not a macro call
type CalledMacro = Macro | string | undefined

export function getCalledMacro(
  callee: NodePath<CallExpression['callee']>,
  importedMacros: Array<ImportedMacro>
): CalledMacro {
  const node = callee.node
  if (isMemberExpression(node)) {
    const ns = node.object
    if (isIdentifier(ns)) {
      if (isLocalDefined(callee.scope, ns.name)) return
      const maybeMacro = importedMacros.find(
        (m) => m.importedAsNamespace && m.local === ns.name
      ) as ImportedAsNS
      if (!maybeMacro) return
      const method = node.property
      if (isIdentifier(method)) {
        return (
          maybeMacro.macros.find((m) => m.name === method.name) || method.name
        )
      } else if (isStringLiteral(method)) {
        // case - namespace['method']()
        return (
          maybeMacro.macros.find((m) => m.name === method.value) || method.value
        )
      }
    }
  } else if (isIdentifier(node)) {
    if (isLocalDefined(callee.scope, node.name)) return
    // case - method()
    const maybeMacro = importedMacros.find(
      (m) => !m.importedAsNamespace && m.local === node.name
    ) as ImportedAsNamed
    if (!maybeMacro) return
    return maybeMacro.macro || node.name
  }
}

export function containsMacros(
  paths: NodePath[],
  importedMacros: ImportedMacro[]
) {
  return paths.map((path) => {
    // traverse can not process the root path
    if (
      path.isCallExpression() &&
      getCalledMacro(path.get('callee'), importedMacros)
    )
      return true
    let has = false
    path.traverse({
      CallExpression(p) {
        if (getCalledMacro(p.get('callee'), importedMacros)) {
          has = true
          p.stop()
        }
      },
    })
    return has
  })
}
