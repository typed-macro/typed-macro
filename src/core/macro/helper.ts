import { ImportOption } from '@/core/helper/import'
import { Node, NodePath } from '@babel/traverse'
import { CallExpression, ImportDeclaration, Program } from '@babel/types'
import {
  normalizePathPattern,
  PathPatternNormalizer,
  projectDir,
} from '@/core/helper/path'
import {
  containsMacros,
  findImported,
  ImportedMacro,
} from '@/core/helper/traverse'
import {
  appendImports,
  appendToBody,
  prependImports,
  prependToBody,
} from '@/core/helper/transform'
import { promise } from '@/common'

export type MacroHelper = {
  /**
   * Prepend import statements to program.
   * @param imports an import or an array of imports
   * @param program node path of the target program for prepending import statements.
   * use the one currently being handled by default.
   * @return node path of the last inserted import statement,
   *   or undefined if imports param is array and length === 0
   */
  prependImports: (
    imports: ImportOption[] | ImportOption,
    program?: NodePath<Program>
  ) => NodePath<ImportDeclaration> | undefined

  /**
   * Append import statements to the last import statement of the program.
   * @param imports an import or an array of imports
   * @param program node path of the target program for prepending import statements.
   * use the one currently being handled by default.
   * @return node path of the last inserted import statement
   *   or undefined if imports param is array and length === 0
   */
  appendImports: (
    imports: ImportOption[] | ImportOption,
    program?: NodePath<Program>
  ) => NodePath<ImportDeclaration> | undefined

  /**
   * Find an import statement that has been in the target program already.
   * Note that findImported() will not resolve extension for moduleName automatically.
   * @param imports an import
   * @param loose match loosely, defaults to true, e.g:
   * `import { a as _a } from 'a'` will match `{ moduleName: 'a' }` and `{ moduleName: 'a', exportName: 'a' }`
   * @param program node path of the target program. use the one currently being handled by default.
   */
  findImported: (
    imports: ImportOption,
    loose?: boolean,
    program?: NodePath<Program>
  ) => NodePath<ImportDeclaration> | undefined

  /**
   * Check if an import statement has been in the target program already.
   * Note that hasImported() will not resolve extension for moduleName automatically.
   * @param imports an import
   * @param loose match loosely, defaults to true, e.g:
   * `import { a as _a } from 'a'` will match `{ moduleName: 'a' }` and `{ moduleName: 'a', exportName: 'a' }`
   * @param program node path of the target program. use the one currently being handled by default.
   */
  hasImported: (
    imports: ImportOption,
    loose?: boolean,
    program?: NodePath<Program>
  ) => boolean

  /**
   * Prepend any node to the target program.
   * @param nodes any node or an array of nodes
   * @param program node path of the target program. use the one currently being handled by default.
   * @return node path of the last inserted node
   */
  prependToBody: <Nodes extends Node | readonly Node[]>(
    nodes: Nodes,
    program?: NodePath<Program>
  ) => NodePath

  /**
   * Append any node to the target program.
   * @param nodes any node or an array of nodes
   * @param program node path of the target program. use the one currently being handled by default.
   * @return node path of the last inserted node
   */
  appendToBody: <Nodes extends Node | readonly Node[]>(
    nodes: Nodes,
    program?: NodePath<Program>
  ) => NodePath

  /**
   * Normalize path pattern so can resolve file paths as import paths
   * @param path relative path or path pattern
   * @param root absolute file path of the project dir, default to projectDir('leaf')
   * @param importer absolute file path of the importer, default to current file
   *
   * e.g. search files by glob pattern and then import them
   *
   * ```typescript
   * // /src/example/a.ts
   * // /src/example/b.ts
   * // in /src/another/index.ts
   * const pattern = '../example/*.ts'
   * const { normalized, base, resolveImportPath } = normalizePathPattern(pattern)
   * const importPaths = glob.sync(normalized, { cwd: base }).map(resolveImportPath)
   *
   * // importPaths = ['../example/a.ts', '../example/b.ts']
   * ```
   */
  normalizePathPattern: (
    pattern: string,
    root?: string,
    importer?: string
  ) => PathPatternNormalizer

  /**
   * Get the directory of the root project or the nearest project.
   * @param which 'root' or 'leaf'
   */
  projectDir: (which: 'root' | 'leaf') => string

  /**
   * Check whether there are unexpanded macros under the path(s).
   * @param path the node path(s) to be checked, use the arguments of the current
   * call expression by default
   */
  containsMacros: (...paths: NodePath[]) => boolean[]

  /**
   * Yield current transformer to nested macros inside current
   * call expression and then call the current macro handler again
   * if the nested macros exist.
   * In another words, expand nested macros in arguments first.
   *
   * It is recommended to call this function at the beginning of the handler
   * that needs to ensure that the arguments don't contain nested macros,
   * to reduce unnecessary repeated calculations.
   *
   * e.g.
   * ```typescript
   * ({ path }, _, { yieldToNestedMacros }) => {
   *   yieldToNestedMacros()
   *   // ...your code
   * }
   * ```
   *
   * Internally, in v0.2.x, it throws a Promise, so DO NOT call it in a try block.
   */
  yieldToNestedMacros: () => never | void
}

export function createHelper(
  thisPath: NodePath<CallExpression>,
  thisProgram: NodePath<Program>,
  thisFilepath: string,
  importedMacros: ImportedMacro[]
) {
  const helpers: MacroHelper = {
    findImported: (imp, loose = true, program = thisProgram) => {
      return findImported(program, imp, loose)
    },

    hasImported: (imp, loose = true, program = thisProgram) => {
      return findImported(program, imp, loose) !== undefined
    },

    prependImports: (imports, program = thisProgram) => {
      return prependImports(program, imports)
    },

    appendImports: (imports, program = thisProgram) => {
      return appendImports(program, imports)
    },

    prependToBody: (nodes, program = thisProgram) => {
      return prependToBody(program, nodes)
    },

    appendToBody: (nodes, program = thisProgram) => {
      return appendToBody(program, nodes)
    },

    normalizePathPattern: (
      pattern,
      root = projectDir('leaf'),
      importer = thisFilepath
    ) => {
      return normalizePathPattern(pattern, importer, root)
    },

    projectDir,

    containsMacros: (...paths) => {
      return containsMacros(
        paths.length ? paths : thisPath.get('arguments'),
        importedMacros
      )
    },

    yieldToNestedMacros: () => {
      if (
        containsMacros(thisPath.get('arguments'), importedMacros).includes(true)
      )
        throw promise
    },
  }

  return helpers
}
