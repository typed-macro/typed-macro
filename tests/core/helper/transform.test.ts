import { getTransformHelper, TransformHelper } from '@/core/helper/transform'
import { File } from '@babel/types'
import { getAST, matchCodeSnapshot } from '../../testutils'
import template from '@babel/template'
import { ImportOption } from '@/core/helper/import'
import traverse from '@babel/traverse'

describe('TransformHelper', () => {
  let ast: File
  let helper: TransformHelper

  const reset = (code = `const z = 1`) => {
    ast = getAST(code)
    helper = getTransformHelper(ast)
  }

  beforeEach(reset)

  it('getProgram() should work', () => {
    expect(helper.getProgram()).not.toBeUndefined()
    traverse(ast, {
      Declaration(path) {
        expect(helper.getProgram(path)).not.toBeUndefined()
        path.stop()
      },
    })
  })

  it('findImported() should return found node path', () => {
    reset(`  
  import { a as _a } from 'a'
  import na from 'a'
  import * as _na from 'a'
  import { a } from 'a'
  import 'a'`)
    const imps: ImportOption[] = [
      { moduleName: 'a', exportName: 'a' },
      { moduleName: 'a', exportName: 'a', localName: '_a' },
      { moduleName: 'a' },
      { moduleName: 'a', defaultName: 'na' },
      { moduleName: 'a', namespaceName: '_na' },
    ]
    imps.forEach((item) =>
      expect(helper.findImported(item)).not.toBeUndefined()
    )
    expect(
      helper.findImported({ moduleName: 'a', exportName: 'b' })
    ).toBeUndefined()
  })

  it('hasImported(loose=true) should work', () => {
    {
      reset(`import { a } from 'a'`)
      expect(helper.hasImported({ moduleName: 'a', exportName: 'a' })).toBe(
        true
      )
      expect(helper.hasImported({ moduleName: 'a' })).toBe(true)
    }
    {
      reset(`import { a as _a } from 'a'`)
      expect(
        helper.hasImported({
          moduleName: 'a',
          exportName: 'a',
          localName: '_a',
        })
      ).toBe(true)
      expect(helper.hasImported({ moduleName: 'a', exportName: 'a' })).toBe(
        true
      )
      expect(helper.hasImported({ moduleName: 'a' })).toBe(true)
    }

    expect(helper.hasImported({ moduleName: 'a', exportName: 'b' })).toBe(false)
  })

  it('hasImported(loose=false) should work', () => {
    {
      reset(`import { a } from 'a'`)
      expect(
        helper.hasImported({ moduleName: 'a', exportName: 'a' }, false)
      ).toBe(true)
      expect(helper.hasImported({ moduleName: 'a' }, false)).toBe(false)
    }
    {
      reset(`import { a as _a } from 'a'`)
      expect(
        helper.hasImported(
          {
            moduleName: 'a',
            exportName: 'a',
            localName: '_a',
          },
          false
        )
      ).toBe(true)
      expect(
        helper.hasImported({ moduleName: 'a', exportName: 'a' }, false)
      ).toBe(false)
      expect(helper.hasImported({ moduleName: 'a' }, false)).toBe(false)
    }

    expect(
      helper.hasImported({ moduleName: 'a', exportName: 'b' }, false)
    ).toBe(false)
  })

  it('pre-/ap-pendImports() should returns undefined if empty array provided', () => {
    reset(`import { a } from 'a'`)
    expect(helper.appendImports([])).toBeUndefined()
    expect(helper.prependImports([])).toBeUndefined()
  })

  it('pre-/ap-pendImports() should returns the existed import stmt', () => {
    reset(`import { a } from 'a'`)
    expect(helper.appendImports({ moduleName: 'a', exportName: 'a' })).toBe(
      helper.findImported({ moduleName: 'a', exportName: 'a' })
    )
    expect(helper.prependImports({ moduleName: 'a', exportName: 'a' })).toBe(
      helper.findImported({ moduleName: 'a', exportName: 'a' })
    )
  })

  it('pre-/ap-pendImports() should insert new import stmts', () => {
    {
      reset(`const a = 1`)
      const imps: ImportOption[] = [
        { moduleName: 'b', exportName: 'b', localName: '_b' },
        { moduleName: 'b', exportName: 'b' },
        { moduleName: 'b' },
        { moduleName: 'b', defaultName: 'nb' },
        { moduleName: 'b', namespaceName: '_nb' },
      ]
      imps.forEach((item) =>
        expect(helper.prependImports(item)).not.toBeUndefined()
      )
      matchCodeSnapshot(ast)
    }
    {
      reset(`const a = 1`)
      const imps: ImportOption[] = [
        { moduleName: 'b', exportName: 'b', localName: '_b' },
        { moduleName: 'b', exportName: 'b' },
        { moduleName: 'b' },
        { moduleName: 'b', defaultName: 'nb' },
        { moduleName: 'b', namespaceName: '_nb' },
      ]
      imps.forEach((item) =>
        expect(helper.appendImports(item)).not.toBeUndefined()
      )
      matchCodeSnapshot(ast)
    }
  })

  it('pre-/ap-pendImports() should unshift to body if no existed import stmt', () => {
    {
      reset(`const b = 1`)
      expect(helper.prependImports({ moduleName: 'a' })).not.toBeUndefined()
      matchCodeSnapshot(ast)
    }
    {
      reset(`const b = 1`)
      expect(helper.appendImports({ moduleName: 'a' })).not.toBeUndefined()
      matchCodeSnapshot(ast)
    }
  })

  it('pre-/ap-pendToBody() should work', () => {
    {
      reset(`const b = 1`)
      helper.prependToBody(template.statement.ast(`console.log('prepended')`))
      matchCodeSnapshot(ast)
    }
    {
      reset(`const b = 1`)
      helper.appendToBody(template.statement.ast(`console.log('appended')`))
      matchCodeSnapshot(ast)
    }
  })
})
