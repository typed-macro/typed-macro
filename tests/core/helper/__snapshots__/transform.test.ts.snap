// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransformHelper pre-/ap-pendImports() should insert new import stmts 1`] = `
"import * as _nb from 'b';
import nb from 'b';
import 'b';
import { b } from 'b';
import { b as _b } from 'b';
const a = 1;"
`;

exports[`TransformHelper pre-/ap-pendImports() should insert new import stmts 2`] = `
"import { b as _b } from 'b';
import { b } from 'b';
import 'b';
import nb from 'b';
import * as _nb from 'b';
const a = 1;"
`;

exports[`TransformHelper pre-/ap-pendImports() should unshift to body if no existed import stmt 1`] = `
"import 'a';
const b = 1;"
`;

exports[`TransformHelper pre-/ap-pendImports() should unshift to body if no existed import stmt 2`] = `
"import 'a';
const b = 1;"
`;

exports[`TransformHelper pre-/ap-pendToBody() should work 1`] = `
"console.log('prepended');
const b = 1;"
`;

exports[`TransformHelper pre-/ap-pendToBody() should work 2`] = `
"const b = 1;
console.log('appended');"
`;
